### ISSUE TITLE
Potential issue with serialization of default values

### ISSUE THREAD

**Issue by cjihrig** [grpc/grpc-node#566]
### Problem description
According to https://developers.google.com/protocol-buffers/docs/proto3#default, default values should not be serialized:

> Also note that if a scalar message field is set to its default, the value will not be serialized on the wire.

However, I am observing default values being serialized.

### Reproduction steps

test.proto:
```
syntax = "proto3";

package test;

service Test {
    rpc Tester (TestRequest) returns (TestResponse) {
    }
}

message TestRequest {
    string str = 1;
    bool boolean = 2;
}

message TestResponse {
}
```

test.js:
```js
'use strict';
const grpc = require('@grpc/grpc-js');
const loader = require('@grpc/proto-loader');
const definition = loader.loadSync('test.proto', {
  keepCase: true,
  longs: String,
  enums: String,
  defaults: true,
  oneofs: true
});
const pkg = grpc.loadPackageDefinition(definition);
const rpc = pkg.test.Test.service.Tester

console.log(rpc.requestSerialize({ str: 'test', boolean: true }));
// <Buffer 0a 04 74 65 73 74 10 01>
console.log(rpc.requestSerialize({ str: 'test', boolean: false }));
// <Buffer 0a 04 74 65 73 74 10 00>
console.log(rpc.requestSerialize({ str: 'test' }));
// <Buffer 0a 04 74 65 73 74>
```

Since `false` is the default Boolean value, I would expect the second and third `Buffer`s to be the same size. Instead, the first and second `Buffer`s are the same size because the default value is serialized.

### Environment
 - OS: macOS
 - Node version: 10.10.0
 - Node installation method: nvm
 - If applicable, compiler version: using pure JS version, so N/A
 - Package name and version: @grpc/grpc-js@0.3.1 and @grpc/proto-loader@0.3.0

### Additional context
Note, if this is a valid issue, it appears to be more of an issue with protobuf, not grpc directly.

**Comment by murgatroid99** [comment]
As you mention, this is a Protobuf.js issue. We don't do any serialization ourselves, we just provide an interface to use Protobuf.js.

**Comment by cjihrig** [comment]
I've opened https://github.com/dcodeIO/protobuf.js/issues/1118. If this gets fixed upstream, I can PR a regression test.

**Comment by nicolasnoble** [comment]
That's... kind of a weird interpretation of the spec, but it's AN interpretation of it. The notion of "default" here is construed as "not set" with protobufjs.

**Comment by geek** [comment]
The reason why this is an issue is because the [gRPC implementation for Go](https://github.com/grpc/grpc-go) follows the proto3 spec and won't serialize default values as part of the message. As a result there is an incompatibility between the two in regards to this basic message generation behavior.

Ideally, there should be compatibility tests setup that ensure that the implementations for the various languages are all creating the same messages.

**Comment by cjihrig** [comment]
FWIW, this isn't purely hypothetical. We ran into actual issues with a Node client talking to Go server. A Node client and Node server worked together fine, and a Go client and Go server worked together fine, but they were doing slightly different things regarding serialization of default values.

**Comment by murgatroid99** [comment]
How would that cause problems? The Go library should still be able to deserialize those messages properly.

**Comment by geek** [comment]
@nicolasnoble the [language guide](https://developers.google.com/protocol-buffers/docs/proto3#default) states the following in regards to default values and serialization:

>  if a scalar message field is set to its default, the value will not be serialized on the wire

**Comment by nicolasnoble** [comment]
I think I'd want to get the protobuf team involved in this one, but can you describe a bit more what you're talking about in terms of calculations on the raw bytes?

The protobuf wire protocol evolved with time, and while it's always been backward compatible, it's definitely possible to have two wire representations that end up with the same serialization, especially between proto2 and proto3 for instance. I'm not sure it's defined behavior of protobuf itself to poke at raw bytes and expecting a strict serialization representation for a given message.

**Comment by murgatroid99** [comment]
I have to add that we have a variety of cross compatibility tests for all of the gRPC languages, and they do all use protobuf as the serialization format. What they test for is API-level compatibility and gRPC protocol compatibility. We do not test for or provide any guarantees about the wire representation being identical.

**Comment by cjihrig** [comment]
I agree that this is a bit of an edge case, but definitely something that caught us off guard due to inconsistencies between languages. I also think this is more of a protobuf issue than grpc, as we've previously agreed.

> especially between proto2 and proto3 for instance

We are sharing the same proto files and proto versions.

> but can you describe a bit more what you're talking about in terms of calculations on the raw bytes?

We're creating a hash of the payload and using it to sign requests (a business requirement out of our control). We're not altering the bytes in any way.

**Comment by nicolasnoble** [comment]
> We're creating a hash of the payload and using it to sign requests (a business requirement out of our control). We're not altering the bytes in any way.

You can't do that.

See https://github.com/protocolbuffers/protobuf/issues/2830 - the serialization output isn't deterministic. While it may be possible to get down to the same length, the content itself isn't guaranteed to be the same. Fields can be serialized in any order, contents of unordered maps can be serialized in any order, depending on the specific memory representation each languages, does, etc.

**Comment by cjihrig** [comment]
That is extremely helpful information. Thank you!

I'm still curious about the serialization behavior of default values, because the text seems pretty straightforward to me, but it suddenly seems much less important 😄

**Comment by murgatroid99** [comment]
Note that that linked issue mentions a "deterministic serialization" API in C++ and Java, but the comments on the C++ header say:

> Note the deterministic serialization is NOT canonical across languages; it is also unstable across different builds with schema changes due to unknown fields.
